core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java

```diff
@@ -0,0 +1,313 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Multimap;
+import io.airlift.units.DataSize;
+import io.airlift.units.Duration;
+import io.trino.client.NodeVersion;
+import io.trino.cost.StatsAndCosts;
+import io.trino.execution.ExecutionFailureInfo;
+import io.trino.execution.NodeTaskMap;
+import io.trino.execution.RemoteTask;
+import io.trino.execution.StageId;
+import io.trino.execution.StateMachine;
+import io.trino.execution.TaskId;
+import io.trino.execution.TaskState;
+import io.trino.execution.TaskStatus;
+import io.trino.execution.buffer.OutputBufferStatus;
+import io.trino.metadata.InMemoryNodeManager;
+import io.trino.metadata.InternalNode;
+import io.trino.metadata.Split;
+import io.trino.sql.planner.Partitioning;
+import io.trino.sql.planner.PartitioningScheme;
+import io.trino.sql.planner.PlanFragment;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.plan.PlanFragmentId;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.testing.TestingMetadata;
+import io.trino.util.FinalizerService;
+import org.testng.annotations.Test;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalLong;
+
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static io.trino.execution.TestingRemoteTaskFactory.TestingRemoteTask;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.sql.planner.SystemPartitioningHandle.SINGLE_DISTRIBUTION;
+import static io.trino.sql.planner.SystemPartitioningHandle.SOURCE_DISTRIBUTION;
+import static io.trino.testing.TestingHandles.TEST_TABLE_HANDLE;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static org.testng.Assert.assertEquals;
+
+public class TestScaledWriterScheduler
+{
+    private static final PlanNodeId TABLE_SCAN_NODE_ID = new PlanNodeId("plan_id");
+    private static final InternalNode NODE_1 = new InternalNode("node-1", URI.create("http://127.0.0.1:11"), new NodeVersion("version-1"), false);
+    private static final InternalNode NODE_2 = new InternalNode("node-2", URI.create("http://127.0.0.1:12"), new NodeVersion("version-1"), false);
+
+    @Test
+    public void testGetNewTaskCountWithUnderutilizedTasksWithoutSkewness()
```

Comment:
I think @gaurav8297 naming suggestion is good, `testScheduleWithUnderutilizedUnskewedTasks`

Reply:
This is the reason why I wanted to test `getNewTaskCount` directly. Unit test scope should not be broader than the function it tests. Otherwise, unit tests can be broken due to changes in codes other than the targeted function. Also, after renaming, it will be hard to know the intention of this unit test.

Reply:
What I and @gaurav8297 mean is that the `testGetNewTaskCountWithUnderutilizedTasksWithoutSkewness` name is simply long. `testScheduleWithUnderutilizedUnskewedTasks` name would convey intend of the test equally well, but is easier to understand.

Reply:
I don't think the length of the test name matters. I think containing the test's exact intention matters. As I said, this test covers `getNewTaskCount` only, not `schedule`.

Reply:
> I don't think the length of the test name matters

But it does for clarity. `getNewTaskCount` is result of `schedule` so both names are technically correct